cns/requestcontroller/kubecontroller/crdreconciler.go

```diff
@@ -30,30 +34,95 @@ func (r *CrdReconciler) Reconcile(request reconcile.Request) (reconcile.Result,
 			return reconcile.Result{}, client.IgnoreNotFound(err)
 		} else {
 			logger.Errorf("[cns-rc] Error retrieving CRD from cache : %v", err)
+			//requeue
 			return reconcile.Result{}, err
 		}
 	}
 
-	logger.Printf("[cns-rc] CRD object: %v", nodeNetConfig)
+	logger.Printf("[cns-rc] CRD spec: %v", nodeNetConfig.Spec)
+	logger.Printf("[cns-rc] CRD status: %v", nodeNetConfig.Status)
 
-	//TODO: Translate CRD status into NetworkContainer request
-	ncRequest, err := CRDStatusToNCRequest(&nodeNetConfig.Status)
+	ipConfigs, err := CRDStatusToCNS(&nodeNetConfig.Status)
 	if err != nil {
 		logger.Errorf("[cns-rc] Error translating crd status to nc request %v", err)
 		//requeue
 		return reconcile.Result{}, err
 	}
 
-	//TODO: process the nc request on CNS side
-	r.CNSClient.UpdateCNSState(ncRequest)
+	if r.CNSClient.ReadyToIPAM() {
+		if err := r.CNSClient.UpdateCNSState(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error updating CNS state: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+	} else {
+		if err := r.markIPs(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error marking ips as allocated when readying CNS: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+
+		if err := r.CNSClient.InitCNSState(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error initializing cns state: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+	}
 
 	return reconcile.Result{}, nil
 }
 
 // SetupWithManager Sets up the reconciler with a new manager, filtering using NodeNetworkConfigFilter
 func (r *CrdReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	//Index nodeNames to later be able to filter by nodeName
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, "spec.nodeName", func(rawObj runtime.Object) []string {
+		pod := rawObj.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&nnc.NodeNetworkConfig{}).
-		WithEventFilter(NodeNetworkConfigFilter{nodeName: r.NodeName}).
+		WithEventFilter(NodeNetworkConfigFilter{nodeName: r.NodeName, namespace: r.Namespace}).
 		Complete(r)
 }
+
+// markIPs marks ips as allocated if it finds a pod using that ip, and available otherwise
+func (r *CrdReconciler) markIPs(ipConfigs []*cns.ContainerIPConfigState) error {
```

Comment:
Thoughts on a different method name, such as `updateIPAvailability`?

Reply:
Yeah sure I'll change that